ksqldb-serde/src/main/java/io/confluent/ksql/serde/avro/AvroSRSchemaDataTranslator.java

```diff
@@ -38,6 +42,89 @@ public class AvroSRSchemaDataTranslator extends ConnectSRSchemaDataTranslator {
     super(schema);
   }
 
+  private List copyArrayData(final List data, final Schema schema) {
+    final List arrayData = new ArrayList();
+
+    for (Object field : data) {
+      if (field instanceof List) {
+        arrayData.add(copyArrayData((List) field, schema.valueSchema()));
+      } else if (field instanceof Map) {
+        arrayData.add(copyMapData((Map) field, schema.valueSchema()));
+      } else if (field instanceof Struct) {
+        arrayData.add(copyStructData((Struct) field, schema.valueSchema()));
+      } else {
+        arrayData.add(field);
+      }
+    }
+
+    return arrayData;
+  }
+
+  private Map copyMapData(final Map<String, Object> data, final Schema schema) {
+    final Map mapData = new HashMap();
+
+    for (Map.Entry<String, Object> entry : data.entrySet()) {
+      final String key = entry.getKey(); // KSQL supports only string keys for Map
+      final Object value = entry.getValue();
+      if (value instanceof List) {
+        mapData.put(key, copyArrayData((List) value, schema.valueSchema()));
+      } else if (value instanceof Map) {
+        mapData.put(key, copyMapData((Map) value, schema.valueSchema()));
+      } else if (value instanceof Struct) {
+        mapData.put(key, copyStructData((Struct) value, schema.valueSchema()));
+      } else {
+        mapData.put(key, value);
+      }
+    }
+
+    return mapData;
+  }
+
+  private Struct copyStructData(final Struct data, final Schema schema) {
+    final Struct innerStruct = new Struct(schema);
+
+    for (Field field : data.schema().fields()) {
+      final Field physicalField = schema.fields().stream()
+          .filter(f -> field.name().equals(f.name())).findFirst().get();
+      final Schema physicalSchema = physicalField.schema();
+
+      if (data.get(field) instanceof List) {
+        innerStruct.put(physicalField, copyArrayData((List) data.get(field), physicalSchema));
+      } else if (data.get(field) instanceof Map) {
+        innerStruct.put(physicalField, copyMapData((Map) data.get(field), physicalSchema));
+      } else if (data.get(field) instanceof Struct) {
+        innerStruct.put(physicalField, copyStructData((Struct) data.get(field), physicalSchema));
+      } else {
+        innerStruct.put(field, data.get(field));
+      }
+    }
+
+    return innerStruct;
+  }
+
+  private void copyDataWithPhysicalSchema(
+      final Struct originalData,
+      final Schema originalSchema,
+      final Struct modifiedData,
+      final Schema schema) {
+    for (final Field field : originalSchema.fields()) {
```

Comment:
QQ: why do we iterate logical schema instead of physical schema. Some fields in physical schema may not appear in logical schema but we want to check if they have defaults or are optional. Basically, what

```
        if (field.schema().defaultValue() != null || field.schema().isOptional()) {
          struct.put(field, field.schema().defaultValue());
        } else {
          throw new KsqlException("Missing default value for required Avro field: [" + field.name()
              + "]. This field appears in Avro schema in Schema Registry");
        }
```
is doing.